source/common/tracing/zipkin_tracer_impl.cc

```diff
@@ -0,0 +1,201 @@
+#include "common/tracing/zipkin_tracer_impl.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/headers.h"
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+#include "common/tracing/http_tracer_impl.h"
+#include "common/tracing/zipkin/zipkin_core_constants.h"
+
+#include "spdlog/spdlog.h"
+
+namespace Tracing {
+
+ZipkinSpan::ZipkinSpan(Zipkin::Span& span) : span_(span) {}
+
+void ZipkinSpan::finishSpan() { span_.finish(); }
+
+void ZipkinSpan::setTag(const std::string& name, const std::string& value) {
+  if (this->hasCSAnnotation()) {
+    span_.setTag(name, value);
+  }
+}
+
+bool ZipkinSpan::hasCSAnnotation() {
+  auto annotations = span_.annotations();
+  if (annotations.size() > 0) {
+    // We currently expect only one annotation to be in the span when this function is called.
+    return annotations[0].value() == Zipkin::ZipkinCoreConstants::get().CLIENT_SEND;
+  }
+  return false;
+}
+
+ZipkinDriver::TlsZipkinTracer::TlsZipkinTracer(Zipkin::Tracer&& tracer, ZipkinDriver& driver)
+    : tracer_(new Zipkin::Tracer(std::move(tracer))), driver_(driver) {}
+
+ZipkinDriver::ZipkinDriver(const Json::Object& config, Upstream::ClusterManager& cluster_manager,
+                           Stats::Store& stats, ThreadLocal::Instance& tls,
+                           Runtime::Loader& runtime, const LocalInfo::LocalInfo& local_info,
+                           Runtime::RandomGenerator& random_generator)
+    : cm_(cluster_manager),
+      tracer_stats_{ZIPKIN_TRACER_STATS(POOL_COUNTER_PREFIX(stats, "tracing.zipkin."))}, tls_(tls),
+      runtime_(runtime), local_info_(local_info), tls_slot_(tls.allocateSlot()) {
+
+  Upstream::ThreadLocalCluster* cluster = cm_.get(config.getString("collector_cluster"));
+  if (!cluster) {
+    throw EnvoyException(fmt::format("{} collector cluster is not defined on cluster manager level",
+                                     config.getString("collector_cluster")));
+  }
+  cluster_ = cluster->info();
+
+  if (cluster_->features() & Upstream::ClusterInfo::Features::HTTP2) {
```

Comment:
Thats a fair point. I put this in there, coz the typical use case is a zipkin server (Java based server) and it doesn't seem to have HTTP/2. I guess, there could be a Envoy there as well.